google/cloud/security/scanner/audit/forwarding_rule_rules_engine.py

```diff
@@ -0,0 +1,253 @@
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book"""
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def: A dictionary containing rule definition properties.
+            rule_index: The index of the rule from the rule definitions.
+            Assigned automatically when the rule book is built.
+
+        Raises:
+
+        """
+
+        target = rule_def.get('target')
+        mode = rule_def.get('mode')
+        load_balancing_scheme = rule_def.get('load_balancing_scheme')
+        port_range = rule_def.get('port_range')
+        port = rule_def.get('port')
+        ip_address = rule_def.get('ip_address')
+        ip_protocol = rule_def.get('ip_protocol')
+        if (target is None) or (mode is None) or \
+           (load_balancing_scheme is None) or \
+           (ip_address is None) or (ip_protocol is None):
+            raise audit_errors.InvalidRulesSchemaError(
+                'Faulty rule {}'.format(rule_def.get('name')))
+        rule_def_resource = {'target': target,
+                             'mode': mode,
+                             'load_balancing_scheme': load_balancing_scheme,
+                             'port_range': port_range,
+                             'ip_address': ip_address,
+                             'ip_protocol': ip_protocol,
+                             'port': port, }
+
+        rule = Rule(rule_name=rule_def.get('name'),
+                    rule_index=rule_index,
+                    rules=rule_def_resource)
+
+        resource_rules = self.resource_rules_map.get(rule_index)
+        if not resource_rules:
+            self.resource_rules_map[rule_index] = rule
+
+    def get_resource_rules(self):
+        """Get all the resource rules for (resource, RuleAppliesTo.*).
+
+        Args:
+            resource: The resource to find in the ResourceRules map.
+
+        Returns:
+            A list of ResourceRules.
+        """
+        resource_rules = []
+
+        for resource_rule in self.resource_rules_map:
+            resource_rules.append(self.resource_rules_map[resource_rule])
+
+        return resource_rules
+
+
+class Rule(object):
+    """Rule properties from the rule definition file.
+    Also finds violations.
+    """
+
+    def __init__(self, rule_name, rule_index, rules):
+        """Initialize.
+
+        Args:
+            rule_name: Name of the loaded rule
+            rule_index: The index of the rule from the rule definitions
+            rules: The rules from the file
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rules = rules
+
+    def find_match(self, forwarding_rule):
+        """Find forwarding rule policy acl violations in the rule book.
```

Comment:
Maybe remove policy from the docs and call it only forwarding rule acl violations or forwarding rule violations